torch/autograd/function.py

```diff
@@ -8,24 +8,54 @@
 from collections import OrderedDict
 from typing import Any, List, Optional
 
+# Formerly known as: _ContextMethodMixin
+# Any class subclassing Function will
+class Ctx(object):
 
-class _ContextMethodMixin(object):
-
-    def save_for_backward(self, *tensors):
+    def save_for_backward(self, *tensors: torch.Tensor):
         r"""Saves given tensors for a future call to :func:`~Function.backward`.
 
         **This should be called at most once, and only from inside the**
-        :func:`forward` **method.**
+        :func:`forward` **method. This should only be called with input or
+        output tensors**
 
-        Later, saved tensors can be accessed through the :attr:`saved_tensors`
+        In :func:`backward`, saved tensors can be accessed through the :attr:`saved_tensors`
         attribute. Before returning them to the user, a check is made to ensure
         they weren't used in any in-place operation that modified their content.
 
         Arguments can also be ``None``.
+
+        See :ref:`extending-autograd` for more details on how to use this method.
+
+        Example::
+            >>> class Func(Function):
+            >>>     @staticmethod
+            >>>     def forward(ctx, x, y, z):
+            >>>         w = x * y * z
+            >>>         out = x * y + y * z + w
+            >>>         ctx.save_for_backward(x, y, out)
+            >>>         ctx.z = z  # z is not a tensor
+            >>>         ctx.w = w  # w is neither input nor output
+            >>>         return out
+            >>>
+            >>>     @staticmethod
+            >>>     def backward(ctx, grad_out):
+            >>>         x, y, out = ctx.saved_tensors
+            >>>         z = ctx.z
+            >>>         gx = grad_out * (y + y * z)
+            >>>         gy = grad_out * (x + z + x * z)
+            >>>         gz = None
+            >>>         return gx, gy, gz
+            >>>
+            >>> a = torch.tensor(1., requires_grad=True, dtype=torch.double)
+            >>> b = torch.tensor(2., requires_grad=True, dtype=torch.double)
+            >>> c = 4
+            >>> d = Func.apply(a, b, c)
+
         """
         self.to_save = tensors
 
-    def mark_dirty(self, *args):
+    def mark_dirty(self, *args: torch.Tensor):
```

Comment:
That would be a tuple of Tensors no?
Same for the other functions

Reply:
I think it would be `tuple[torch.Tensor, ...]` if we expect a single argument `args`, but for `*args`, we should specify what we expect a single argument should be:
https://stackoverflow.com/questions/48909673/how-to-annotate-variadic-parameters-in-python-using-typing-annotations

Reply:
Ho that's surprising. ok!